sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -321,6 +321,38 @@ object GeneratorNestedColumnAliasing {
     // need to prune nested columns through Project and under Generate. The difference is
     // when `nestedSchemaPruningEnabled` is on, nested columns will be pruned further at
     // file format readers if it is supported.
+
+    // There are [[ExtractValue]] expressions on or not on the output of the generator. Generator
+    // can also have different types:
+    // 1. For [[ExtractValue]]s not on the output of the generator, theoretically speaking, there
+    //    lots of expressions that we can push down, including non ExtractValues and GetArrayItem
+    //    and GetMapValue. But to be safe, we only handle GetStructField and GetArrayStructFields.
```

Comment:
The first item looks a bit weird. For what `ExtractValue` can be pushed down, you can simply list them. I'm not sure "For [[ExtractValue]]s not on the output of the generator..." means. Do you mean on the output of generator?

Reply:
```
val (attrToExtractValuesOnGenerator, attrToExtractValuesNotOnGenerator) =
  attrToExtractValues.partition { case (attr, _) =>
    attr.references.subsetOf(generatorOutputSet) }

val pushedThrough = NestedColumnAliasing.rewritePlanWithAliases(
  plan, attrToExtractValuesNotOnGenerator)
```

We call `NestedColumnAliasing.rewritePlanWithAliases` for `ExtractValue`s not on the output of the generators. The above code snippet is also the reason that I was hesitated to modify `canPruneGenerator`. If we modify `canPruneGenerator`, we miss the opportunity for `Inline` and `Stack` to use `NestedColumnAliasing.rewritePlanWithAliases` on `attrToExtractValuesNotOnGenerator`. My point in the comment was to note the difference of optimizations able to apply to `attrToExtractValuesOnGenerator` and `attrToExtractValuesNotOnGenerator`. The former one is predicate pushdown and there are lots of expressions that can be pushed down, the latter one is the real "GeneratorNestedColumnAliasing" and only `GetStructField` is supported.